### DISCUSSION TITLE
Autenticazione con tado tramite Oauth2

### DISCUSSION THREAD

**Discussion by Newreader** [indomus/forum#2861] [GENERALE - Discussioni generiche]
Salve
Ho ricevuto la mail sotto riportata da tado:
Grazie per utilizzare tado! 
Abbiamo notato una serie di richieste obsolete di dati domestici dal tuo account al cloud tado.
È improbabile che queste richieste siano state generate dalle versioni attuali delle app o dei dispositivi tado.
Se hai configurato un sistema domotico open source o commerciale non elencato nella nostra pagina di integrazione della smart home, o stai ancora utilizzando una versione molto vecchia dell'app tado, segui le istruzioni alla fine dell'e-mail.
Queste particolari richieste ai nostri server tado utilizzano una forma di autenticazione obsoleta e non sicura.
(In termini tecnici: queste richieste aggiungono nome utente e password come parametri di query a ciascuna richiesta, ad esempio
[https://my.tado.com/api/v2/me?username=myUsername&password=myPassword]
Tieni presente che questa forma obsoleta di autenticazione presto smetterà di funzionare.
Prevediamo di disattivarla il 12 febbraio 2024.
Cosa ti chiediamo di fare?
-Se utilizzi un software di domotica che si connette al cloud tado, aggiornalo alla versione più recente o chiedi al fornitore del software una versione del loro software di autenticazione con tado tramite Oauth2
-Se hai scritto tu stesso il software, puoi trovare informazioni su come autenticare correttamente le tue richieste [in questa documentazione].
-Assicurati anche di utilizzare il file della versione attuale dell'app tado. Puoi aggiornare l'app su [Google Play Store] o sull’[Apple App Store].
Se il tuo dispositivo mobile non è più supportato, puoi anche fare riferimento alla [web-app tado].
Cordiali saluti
Il team di tado

Premetto che in casa ho installato da qualche mese il Termostato Intelligente Cablato V3+ e alcune teste termostatiche intelligenti V3+.
Utilizzo sia l’app originale tado scaricata dal Play Store che l’integrazione tado di Home Assistant (Core con Docker installato su un Raspberry Pi 3B+ aggiornato alla versione 2024.1.3), che penso già utilizzi l’autenticazione Oauth2 o sbaglio?
In aggiunta ho qualche automazione che mi regola l’offset di ciascuna testa.
Finora tutto ha funzionato egregiamente senza alcun problema
Tutto ciò premesso vorrei capire se qualcun altro ha ricevuto una comunicazione simile e soprattutto cosa devo fare per soddisfare le richieste (che secondo me, non trovandomi nelle condizioni ipotizzate nella mail, sono già tutte soddisfatte) di autenticazione con tado tramite Oauth2.
Grazie

**Comment by snakuzzo** [comment]
Io non ancora, ma ho letto che è arrivata a diverse persone.
Non vorrei sbagliarmi, ma di fatto ora l'integrazione su Core non dovrebbe funzionare con oauth2...ho visto al volo il codice e mi sa che usa solo user e password.
Io ho creato un mio package per regolare l'offset delle valvole, e lì invece utilizzo oauth2:

- Si fa prima auth con POST verso auth.tado.com passando nel body username, password, client id, client secret e grant type
- Si recupera quindi l'access token dalla response
- Si fa la chiamata in GET verso my.tado.com/api/v2/me passando nell'header authorization il bearer token (con il token recuperato prima appunto)

![image](https://github.com/indomus/forum/assets/7205371/2476f1a4-b810-4df9-b3d3-bf9ef83d5b0f)

![image](https://github.com/indomus/forum/assets/7205371/4e1d574f-927c-4c91-869f-8cd658593ad6)

**Reply by snakuzzo** [reply]
Fatto un'ulteriore verifica. 
In realtà l'integrazione utilizza [PyTado](https://github.com/wmalgadey/PyTado) che già prevede il giro corretto.
Di seguito il link diretto alla funzione di login...

https://github.com/wmalgadey/PyTado/blob/863b9633747b4aa6fe2e0afaed1481a4364192a2/PyTado/http/http.py#L224

**Reply by Newreader** [reply]
Grazie per la risposta.
Mi sembra di capire che l'integrazione tado su HA utilizza PYTado che prevede già la forma di autenticazione richiesta partendo dalla username e password che si inseriscono la prima volta che si avvia l'intergazione. Quindi non occorre fare alcunché. Giusto?
Pur tuttavia come mai tado mi scrive "Queste particolari richieste ai nostri server tado utilizzano una forma di autenticazione obsoleta e non sicura. (In termini tecnici: queste richieste aggiungono nome utente e password come parametri di query a ciascuna richiesta."?
La procedura che tu mi hai descritto la ritrovo anche qui:

https://support.tado.com/en/articles/8565472-how-do-i-update-my-rest-api-authentication-method-to-oauth-2

**Reply by snakuzzo** [reply]
Non ho visto l'integrazione com'è fatta...ma in teoria non credo sarà necessario fare qualche modifica lato utente.

**Reply by snakuzzo** [reply]
Ho dato un'occhiata più approfondita. E' già tutto come deve essere...

L'integrazione Tado di Core utilizza il modulo presente su pypi.org python-tado (https://pypi.org/project/python-tado/#description) che afferisce al progetto PyTado su github (https://github.com/wmalgadey/PyTado/releases) e nello specifico l'ultima versione di Core rilasciata utilizza l'ultima versione del modulo 0.17.4 nel cui codice c'è già il giro di auth nuovo (per modo di dire, perché io lo utilizzo già da almeno 2 anni). Questo è il permalink della funzione https://github.com/wmalgadey/PyTado/blob/863b9633747b4aa6fe2e0afaed1481a4364192a2/PyTado/http/http.py#L224

Qui l'evdenza del codice

![image](https://github.com/indomus/forum/assets/7205371/3f5f84ab-34d5-48a4-8af2-ee96235f5022)

Qui la dimostrazione della versione del modulo utilizzato da Core

![image](https://github.com/indomus/forum/assets/7205371/5f687bdc-29df-4ee3-aa06-985623aaa7cc)

Qui invece, come puoi vedere, viene già utilizzata la URL corretta per la login e la refresh token...

![image](https://github.com/indomus/forum/assets/7205371/2d5732a8-c834-4aee-80e4-fb97fc3e76e5)

**Reply by Newreader** [reply]
Grazie per aver dato la dimostrazione tecnica delle mie asserzioni.
